### ISSUE TITLE
why use tools/test.py and mmdet.apis to inference  a same checkpoint with same cfg file but generate different output

### ISSUE THREAD

**Issue by lisenjie757** [open-mmlab/mmdetection#10395]
## code using mmdet.apis to inference

```py
from mmdet.apis import init_detector, inference_detector, async_inference_detector
from mmdet.utils import register_all_modules
import torch
import torchvision.transforms as transforms
from PIL import Image
import mmcv
import numpy as np
import os
import glob
import xmltodict
from mmdet.datasets.voc import VOCDataset
from mmengine.fileio import list_from_file

config_file = 'configs/yolo/yolov3_mobilenetv2_8xb24-ms-416-300e_coco.py'
checkpoint_file = 'work_dirs/yolov3_mobilenetv2_pretrained/best_pascal_voc_mAP_epoch_27.pth'
class_dict = {
    'aeroplane': 0,
    'bicycle': 1,
    'bird': 2,
    'boat': 3,
    'bottle': 4,
    'bus': 5,
    'car': 6,
    'cat': 7,
    'chair': 8,
    'cow': 9,
    'diningtable': 10,
    'dog': 11,
    'horse': 12,
    'motorbike': 13,
    'person': 14,
    'pottedplant': 15,
    'sheep': 16,
    'sofa': 17,
    'train': 18,
    'tvmonitor': 19
}

#Register all modules in mmdet into the registries
register_all_modules()

device = torch.device("cpu")

model = init_detector(config_file, checkpoint_file, device=device)
model.eval()

test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', scale=(224, 224), keep_ratio=True),
    # avoid bboxes being resized
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='PackDetInputs',
        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
                   'scale_factor'))
]

data = VOCDataset(ann_file='../data/VOCdevkit/VOC2007/ImageSets/Main/test.txt',
        img_subdir='../data/VOCdevkit/VOC2007/test_img/',
        ann_subdir='../data/VOCdevkit/VOC2007/Annotations/',
        test_mode=True,
        pipeline=test_pipeline)

img_ids = list_from_file("../data/VOCdevkit/VOC2007/ImageSets/Main/test.txt")

annotations, det_results  = [], []
for i in range(len(data)):

    ann = {'bboxes': data[i]['data_samples'].gt_instances.bboxes.cpu().numpy(),
           'labels': data[i]['data_samples'].gt_instances.labels.cpu().numpy(),
           'bboxes_ignore': data[i]['data_samples'].ignored_instances.bboxes.cpu().numpy(),
           'labels_ignore': data[i]['data_samples'].ignored_instances.labels.cpu().numpy()}
    annotations.append(ann)

    img_path = os.path.join('../data/VOCdevkit/VOC2007/JPEGImages/', img_ids[i] + '.jpg')

    img = mmcv.imread(img_path, channel_order='rgb')
    output2 = inference_detector(model, img)

    pred_bboxes = output2.pred_instances.bboxes.cpu().numpy()
    pred_scores = output2.pred_instances.scores.cpu().numpy()
    pred_labels = output2.pred_instances.labels.cpu().numpy()
    dets = []
    for label in range(len(class_dict)):
        index = np.where(pred_labels == label)[0]
        pred_bbox_scores = np.hstack(
            [pred_bboxes[index], pred_scores[index].reshape((-1, 1))])
        dets.append(pred_bbox_scores)
    
    det_results.append(dets)

from mmdet.evaluation.functional.mean_ap import eval_map

mean_ap, eval_results = eval_map(det_results, annotations, eval_mode='11points', use_legacy_coordinate=True)
```

## output
```
+-------+------+------+--------+-------+
| class | gts  | dets | recall | ap    |
+-------+------+------+--------+-------+
| 0     | 311  | 705  | 0.537  | 0.475 |
| 1     | 389  | 1842 | 0.658  | 0.537 |
| 2     | 576  | 720  | 0.398  | 0.321 |
| 3     | 393  | 1654 | 0.364  | 0.233 |
| 4     | 657  | 1205 | 0.189  | 0.157 |
| 5     | 254  | 712  | 0.626  | 0.514 |
| 6     | 1541 | 2840 | 0.517  | 0.466 |
| 7     | 370  | 773  | 0.676  | 0.581 |
| 8     | 1374 | 3625 | 0.370  | 0.223 |
| 9     | 329  | 477  | 0.422  | 0.325 |
| 10    | 299  | 1486 | 0.569  | 0.380 |
| 11    | 530  | 1325 | 0.668  | 0.533 |
| 12    | 395  | 899  | 0.646  | 0.529 |
| 13    | 369  | 1460 | 0.726  | 0.620 |
| 14    | 5227 | 8503 | 0.550  | 0.467 |
| 15    | 592  | 1324 | 0.385  | 0.278 |
| 16    | 311  | 511  | 0.447  | 0.359 |
| 17    | 396  | 1156 | 0.624  | 0.428 |
| 18    | 302  | 750  | 0.682  | 0.587 |
| 19    | 361  | 1147 | 0.565  | 0.427 |
+-------+------+------+--------+-------+
| mAP   |      |      |        | 0.422 |
+-------+------+------+--------+-------+
```

## use tools/test.py

```shell
python tools/test.py configs/yolo/yolov3_mobilenetv2_8xb24-ms-416-300e_coco.py work_dirs/yolov3_mobilenetv2_pretrained/best_pascal_voc_mAP_epoch_27.pth
```
## output
```
+-------------+------+------+--------+-------+
| class       | gts  | dets | recall | ap    |
+-------------+------+------+--------+-------+
| aeroplane   | 311  | 732  | 0.598  | 0.531 |
| bicycle     | 389  | 1567 | 0.674  | 0.567 |
| bird        | 576  | 800  | 0.460  | 0.390 |
| boat        | 393  | 1213 | 0.372  | 0.297 |
| bottle      | 657  | 1341 | 0.250  | 0.185 |
| bus         | 254  | 767  | 0.669  | 0.555 |
| car         | 1541 | 2760 | 0.566  | 0.506 |
| cat         | 370  | 733  | 0.754  | 0.665 |
| chair       | 1374 | 4081 | 0.424  | 0.266 |
| cow         | 329  | 774  | 0.574  | 0.437 |
| diningtable | 299  | 1582 | 0.679  | 0.488 |
| dog         | 530  | 1413 | 0.774  | 0.630 |
| horse       | 395  | 1165 | 0.754  | 0.654 |
| motorbike   | 369  | 1383 | 0.751  | 0.643 |
| person      | 5227 | 9672 | 0.622  | 0.556 |
| pottedplant | 592  | 1703 | 0.444  | 0.324 |
| sheep       | 311  | 685  | 0.556  | 0.434 |
| sofa        | 396  | 1157 | 0.669  | 0.451 |
| train       | 302  | 766  | 0.772  | 0.665 |
| tvmonitor   | 361  | 1023 | 0.637  | 0.542 |
+-------------+------+------+--------+-------+
| mAP         |      |      |        | 0.489 |
+-------------+------+------+--------+-------+
```

## my cfg file

i.e. configs/yolo/yolov3_mobilenetv2_8xb24-ms-416-300e_coco.py

```py
train_cfg = dict(type='EpochBasedTrainLoop', max_epochs=200, val_interval=1)
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')
param_scheduler = [
    dict(
        type='LinearLR',
        start_factor=0.0001,
        by_epoch=False,
        begin=0,
        end=4000),
    dict(type='MultiStepLR', by_epoch=True, milestones=[24, 28], gamma=0.1)
]
optim_wrapper = dict(
    type='AmpOptimWrapper',
    optimizer=dict(type='SGD', lr=0.003, momentum=0.9, weight_decay=0.0005),
    clip_grad=dict(max_norm=35, norm_type=2),
    loss_scale='dynamic')
auto_scale_lr = dict(enable=True, base_batch_size=128)
default_scope = 'mmdet'
default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=50),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(
        type='CheckpointHook', interval=1, save_best='auto', max_keep_ckpts=2),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    visualization=dict(type='DetVisualizationHook'))
env_cfg = dict(
    cudnn_benchmark=False,
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    dist_cfg=dict(backend='nccl'))
vis_backends = [dict(type='LocalVisBackend'), dict(type='WandbVisBackend')]
visualizer = dict(
    type='DetLocalVisualizer',
    vis_backends=[dict(type='LocalVisBackend'),
                  dict(type='WandbVisBackend')],
    name='visualizer')
log_processor = dict(type='LogProcessor', window_size=50, by_epoch=True)
visualization = dict(
    type='DetVisualizationHook', draw=True, interval=1, show=True)
log_level = 'INFO'
load_from = None
resume = False
data_preprocessor = dict(
    type='DetDataPreprocessor',
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    bgr_to_rgb=True,
    pad_size_divisor=32)
model = dict(
    type='YOLOV3',
    data_preprocessor=dict(
        type='DetDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True,
        pad_size_divisor=32),
    backbone=dict(
        type='MobileNetV2',
        out_indices=(2, 4, 6),
        act_cfg=dict(type='LeakyReLU', negative_slope=0.1),
        init_cfg=dict(
            type='Pretrained', checkpoint='open-mmlab://mmdet/mobilenet_v2')),
    neck=dict(
        type='YOLOV3Neck',
        num_scales=3,
        in_channels=[320, 96, 32],
        out_channels=[96, 96, 96]),
    bbox_head=dict(
        type='YOLOV3Head',
        num_classes=80,
        in_channels=[96, 96, 96],
        out_channels=[96, 96, 96],
        anchor_generator=dict(
            type='YOLOAnchorGenerator',
            base_sizes=[[(116, 90), (156, 198), (373, 326)],
                        [(30, 61), (62, 45), (59, 119)],
                        [(10, 13), (16, 30), (33, 23)]],
            strides=[32, 16, 8]),
        bbox_coder=dict(type='YOLOBBoxCoder'),
        featmap_strides=[32, 16, 8],
        loss_cls=dict(
            type='CrossEntropyLoss',
            use_sigmoid=True,
            loss_weight=1.0,
            reduction='sum'),
        loss_conf=dict(
            type='CrossEntropyLoss',
            use_sigmoid=True,
            loss_weight=1.0,
            reduction='sum'),
        loss_xy=dict(
            type='CrossEntropyLoss',
            use_sigmoid=True,
            loss_weight=2.0,
            reduction='sum'),
        loss_wh=dict(type='MSELoss', loss_weight=2.0, reduction='sum')),
    train_cfg=dict(
        assigner=dict(
            type='GridAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.5,
            min_pos_iou=0)),
    test_cfg=dict(
        nms_pre=1000,
        min_bbox_size=0,
        score_thr=0.05,
        conf_thr=0.005,
        nms=dict(type='nms', iou_threshold=0.45),
        max_per_img=100))
dataset_type = 'VOCDataset'
data_root = '../data/VOCdevkit/'
backend_args = None
train_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='RandomResize', scale=[(224, 224), (416, 416)], keep_ratio=True),
    dict(type='RandomFlip', prob=0.5),
    dict(type='PackDetInputs')
]
test_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(type='Resize', scale=(224, 224), keep_ratio=True),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='PackDetInputs',
        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
                   'scale_factor'))
]
train_dataloader = dict(
    batch_size=64,
    num_workers=15,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    batch_sampler=dict(type='AspectRatioBatchSampler'),
    dataset=dict(
        type='RepeatDataset',
        times=3,
        dataset=dict(
            type='ConcatDataset',
            ignore_keys=['dataset_type'],
            datasets=[
                dict(
                    type='VOCDataset',
                    data_root='../data/VOCdevkit/',
                    ann_file='VOC2007/ImageSets/Main/trainval.txt',
                    data_prefix=dict(sub_data_root='VOC2007/'),
                    filter_cfg=dict(
                        filter_empty_gt=True, min_size=32, bbox_min_size=32),
                    pipeline=[
                        dict(type='LoadImageFromFile', backend_args=None),
                        dict(type='LoadAnnotations', with_bbox=True),
                        dict(
                            type='RandomResize',
                            scale=[(224, 224), (416, 416)],
                            keep_ratio=True),
                        dict(type='RandomFlip', prob=0.5),
                        dict(type='PackDetInputs')
                    ],
                    backend_args=None),
                dict(
                    type='VOCDataset',
                    data_root='../data/VOCdevkit/',
                    ann_file='VOC2012/ImageSets/Main/trainval.txt',
                    data_prefix=dict(sub_data_root='VOC2012/'),
                    filter_cfg=dict(
                        filter_empty_gt=True, min_size=32, bbox_min_size=32),
                    pipeline=[
                        dict(type='LoadImageFromFile', backend_args=None),
                        dict(type='LoadAnnotations', with_bbox=True),
                        dict(
                            type='RandomResize',
                            scale=[(224, 224), (416, 416)],
                            keep_ratio=True),
                        dict(type='RandomFlip', prob=0.5),
                        dict(type='PackDetInputs')
                    ],
                    backend_args=None)
            ])))
val_dataloader = dict(
    batch_size=64,
    num_workers=15,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='VOCDataset',
        data_root='../data/VOCdevkit/',
        ann_file='VOC2007/ImageSets/Main/test.txt',
        data_prefix=dict(sub_data_root='VOC2007/'),
        test_mode=True,
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(type='Resize', scale=(224, 224), keep_ratio=True),
            dict(type='LoadAnnotations', with_bbox=True),
            dict(
                type='PackDetInputs',
                meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
                           'scale_factor'))
        ],
        backend_args=None))
test_dataloader = dict(
    batch_size=64,
    num_workers=15,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='VOCDataset',
        data_root='../data/VOCdevkit/',
        ann_file='VOC2007/ImageSets/Main/test.txt',
        data_prefix=dict(sub_data_root='VOC2007/'),
        test_mode=True,
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(type='Resize', scale=(224, 224), keep_ratio=True),
            dict(type='LoadAnnotations', with_bbox=True),
            dict(
                type='PackDetInputs',
                meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
                           'scale_factor'))
        ],
        backend_args=None))
val_evaluator = dict(type='VOCMetric', metric='mAP', eval_mode='11points')
test_evaluator = dict(type='VOCMetric', metric='mAP', eval_mode='11points')
find_unused_parameters = True
launcher = 'pytorch'
work_dir = './work_dirs/yolov3_mobilenetv2_8xb24-ms-416-300e_coco'
```

**Comment by lisenjie757** [comment]
please help me， this is import to me

**Comment by okt7760** [comment]
@lisenjie757 Why do I get a json file when I'm reasoning with test.py and you get evaluation output

**Comment by lisenjie757** [comment]
@okt7760 i use mmdetction3.0，this is the terminal output.

> @lisenjie757 Why do I get a json file when I'm reasoning with test.py and you get evaluation output

**Comment by lisenjie757** [comment]
@okt7760  and if you solve your problem, please help me check if test.py and mmedet.apis have the same output in your checkpoint and cfg file, i would be very grateful to you.

> @lisenjie757 Why do I get a json file when I'm reasoning with test.py and you get evaluation output

**Comment by okt7760** [comment]
I used the coco type data set, while I also used mmdetection3.0, my evaluator configuration is as follows, have you used the coco type for testing

=================================================
val_evaluator = dict(
    type='CocoMetric',
    ann_file=r"E:\data\visdrone\visdrone_json\val.json",
    metric='bbox',
    format_only=False,
    backend_args=backend_args)
#test_evaluator = val_evaluator

# inference on test dataset and
# format the output results for submission.

test_evaluator = dict(
    type='CocoMetric',
    metric='bbox',
    format_only=True,
    ann_file=r"E:\data\visdrone\visdrone_json\test.json",
    outfile_prefix='./work_dirs/coco_detection/test')

**Comment by lisenjie757** [comment]
@okt7760 yes, i have tried it, it works fine. i think you can try default config file used coco type evaluator to have a test.

**Comment by okt7760** [comment]
@lisenjie757 Thanks i found it eval only generates json files when my format_only is True. I changed it to false its work

**Comment by hy0523** [comment]
@lisenjie757 Hello, have you solve this problem? I meet the same phenomenon, can you give some suggestions? Thanks.
